pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -198,54 +198,56 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (result re
 		return result, err
 	}
 
-	// Check for a "discovered" host vs. one that we have all the info for.
-	if host.Spec.BMC.Address == "" {
-		reqLogger.Info(bmc.MissingAddressMsg)
-		dirty := host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusDiscovered)
-		if dirty {
-			err = r.saveStatus(host)
-			if err != nil {
+	// Retrieve the BMC details from the host spec and validate host
+	// BMC details and build the credentials for talking to the
+	// management controller.
+	bmcCreds, bmcCredsSecret, err := r.buildAndValidateBMCCredentials(request, host)
+	if err != nil {
+		switch err.(type) {
+
+		// We key on invalid bmc address and secret references as a trigger
+		// the host needs to be put into a discovered status and requeue it for
+		// later.  We do set an error message (but not an error state) on the host
+		// so we can understand what we may be waiting on.
+		case *InvalidBMCAddressError, *InvalidBMCSecretError:
```

Comment:
We need make sure we continue to differentiate between missing values and invalid values here.

Missing values will come about when we eventually get the "discovery" feature working by polling ironic to find hosts that we don't already know about. Those won't have BMC credentials at all, so we will create host objects with those fields empty.

Invalid values are any non-empty values that do not work when we try to talk to the BMC. When we get those, we should put the host into the registration error state rather than the discovered state.

It looks like InvalidBMCAddressError could be renamed MissingBMCAddressError since that's the only way it's used, but InvalidBMCSecretError is used both for a missing secret (discovered host) and for a secret with a missing value (registration error), so I think we need to error types there.

Reply:
@dhellmann I took a stab at refactoring the types so they were a bit more human readable.  This was pretty simple with the new structure so thats a good sign.  I hope we've got it where it should be now.